src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/TestUtils.kt

```diff
@@ -55,53 +54,88 @@ private fun compareAndOverwrite(expectedOutputFilePrefix: String, actualOutput:
+            val overwriteOutputFileName = generateExpectedLogFileName(
+                fileNamePrefix = expectedOutputFilePrefix,
+                jdkVersion = testJdkVersion,
+                inTraceDebuggerMode = isInTraceDebuggerMode,
+            )
+            val overwriteOutputFile = getExpectedLogFileFromSources(overwriteOutputFileName)
+            overwriteOutputFile.writeText(actualOutput)
         } else {
             assertEquals(expectedOutput, actualOutput)
         }
     }
 }
 
-// To prevent file duplication this function finds the file to compare the results to.
-// With preference for jdk: 17, 15, 13, 11, 8 (trace mode) and 17, 15, 13, 11, 8 (non trace mode)
-// Search starts at current jdk level
-// For instance we are running tests for jdk 11 (trace mode), we will check file existence in the following order:
-// 11 (trace), 8 (trace), 11 (non-trace) and 8 (non-trace).
-private fun getFileToCompareTo(expectedOutputFilePrefix: String): String {
-    // If in trace mode first check if a trace debugger file can be found
+/* To prevent file duplication, this function finds the file to compare the results to.
+ * With preference for jdk: 17, 15, 13, 11, 8 (trace mode) and 17, 15, 13, 11, 8 (non-trace mode).
+ * Search starts at the current jdk level.
+ * For instance, we are running tests for jdk 11 (trace mode),
+ * we will check file existence in the following order:
+ * 11 (trace), 8 (trace), 11 (non-trace) and 8 (non-trace).
+ */
+private fun getExpectedLogFile(expectedOutputFilePrefix: String): File {
+    // first try to pick a jdk-specific file if it exists
+    val jdkSpecificFileName = generateExpectedLogFileName(
+        fileNamePrefix = expectedOutputFilePrefix,
+        jdkVersion = testJdkVersion,
+        inTraceDebuggerMode = isInTraceDebuggerMode,
+    )
+    val jdkSpecificFile = getExpectedLogFileFromResources(jdkSpecificFileName)
+    if (jdkSpecificFile != null) return jdkSpecificFile
+
+    // next, if in trace debugger mode, try a non-jdk-specific file with trace debugger suffix
     if (isInTraceDebuggerMode) {
-        for (i in testJdkVersion.ordinal downTo 0) {
-            val jdkVersion = TestJdkVersion.entries[i]
-            val fileName = getFileNameFor(expectedOutputFilePrefix, jdkVersion, true)
-            if (logFileFromResourcesExists(fileName)) return fileName
-        }
-    }
-    // Check lower sdks in non-trace mode
-    for (i in testJdkVersion.ordinal downTo 0) {
-        val jdkVersion = TestJdkVersion.entries[i]
-        val fileName = getFileNameFor(expectedOutputFilePrefix, jdkVersion, false)
-        if (logFileFromResourcesExists(fileName)) return fileName
+        val traceDebuggerFileName = generateExpectedLogFileName(
+            fileNamePrefix = expectedOutputFilePrefix,
+            jdkVersion = DEFAULT_TEST_JDK_VERSION,
+            inTraceDebuggerMode = true,
```

Comment:
Wouldn't it make more sense to compare to `jdkVersion = testJdkVersion` and `inTraceDebuggerMode = false` aka compare trace tot non-trace of same jdk.

Often the output is the same for trace and non-trace modes, but when there are a lot of different outputs on the jdk axis (like the `thread_pool_jdkN.txt` files) we would have a lot of duplicates the for the trace mode.

Reply:
I actually implemented both approaches, please have a look : )